rfcs/0052-hard-fork-data-migration.md

```diff
@@ -0,0 +1,145 @@
+ - create a new, empty database using the new archive schema
+ - run the first migration app against the mainnet and new databases
+ - run the second migration app with the `--checkpoint-interval` set
+    to some suitable value (perhaps 100), and starting with the
+	original mainnet ledger in the input file
+ - use `pg_dump` to dump the migrated database, upload it
+ - upload the most recent checkpoint file
+
+The cron job will perform these same steps in an automated fashion:
+ - pull latest mainnet archive dump, load into PostgresQL
+ - pull latest migrated database, load into PostgreSQL
+ - pull latest checkpoint file
+ - run first migration app against the two databases
+ - run second migration app, using the downloaded checkpoint file; checkpoint interval
+    should be smaller (perhaps 50), because there are typically only 200 or so blocks in a day
+ - upload migrated database
+ - upload most recent checkpoint file
+
+There should be monitoring of the cron job, in case there are errors.
+
+Just before the hard fork, the last few blocks can be migrated by running locally:
+- download the mainnet archive data directly from the k8s PostgreSQL node, not from
+  the archive dump, load it into PostgreSQL
+- download the most recent migrated database, load it into PostgresQL
+- download the most recent checkpoint file
+- run the first migration application against the two database
+- run the second migration application using the most recent checkpoint file
+
+It is worthwhile to perform these last steps as a `dry run` to make sure all goes
+well. Those steps can be run as many times as needed.
+
+## Drawbacks
+[drawbacks]: #drawbacks
+
+If we want mainnet data to be available after the hard fork, there
+needs to be migration of that data.
+
+## Rationale and alternatives
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+It may be possible to add or delete columns in the original schema to
+perform some of the migration without transferring data between
+databases. It would still be necessary to add the windowing data from
+precomputed blocks, and to have a separate pass to populate the
+`accounts...` tables.
+
+## Prior art
+[prior-art]: #prior-art
+
+There are preliminary implementations of the two applications:
+
+- The first application is in branch `feature/berkeley-db-migrator`.
+Downloading precomputed blocks appears to be the main bottleneck there,
+so those blocks are downloaded in batches, which helps considerably.
+
+- The second application is in branch `feature/add-berkeley-accounts-tables`.
+
+There has been some local testing of these applications.
+
+The replayer cron jobs for mainnet, devnet, and berkeley can serve as
+a starting point for the implementation of the cron job described
+here.
+
+## Unresolved questions
+[unresolved-questions]: #unresolved-questions
+
+How do we limit the migration to the final block of mainnet? There could be
+flags to the migration apps to stop at a given state hash or height.
+
```

Comment:
OK, in the first-phase migration app, in #12906, I've add a `--end-global-slot` command-line arg, and tested that. The second-phase app doesn't need to worry about the end slot, because it can only process what the first-phase app has produced.

If you omit that arg, the app will migrate only canonical blocks. My understanding is that there may be some pending blocks to be migration, so if you do provide that arg, the app will migrate both pending and canonical (but not orphaned) blocks.

Reply:
I removed the unresolved question, because I think this command-line arg solves the problem.